engine/ingame/controller.py

```diff
@@ -53,16 +53,28 @@ def on_mouse_release(
             self.model.activate()
 
     def on_update(self, delta_time: int) -> None:
+        """
+        Updates player speed based on keypress,
+        and prevents player going out-of-bounds.
+        """
         vx, vy = 0, 0
 
-        if self._up_key_pressed():
+        player_sprite = self.model.player_sprite
+        map_width = self.model.width * self.model.tile_width
+        map_height = self.model.height * self.model.tile_height
+
+        if (self._up_key_pressed() and not
+        (player_sprite.top >= map_height)):
```

Comment:
Recommendation: set up your editor to automatically run the black formatter on save.